File: ts/packages/protocol/src/client-core.ts
import { randomUUID } from 'node:crypto';
import type { Transport } from './transport.js';
import { SERVICE_RE, METHOD_RE, parseEnvelope, EnvelopeKind, buildRequest, buildNotification } from './envelope.js';
import { RpcError, ClamatorProtocolError } from './error.js';

export interface ClamatorClient {
  call<P, R>(service: string, method: string, params: P): Promise<R>;
  notify<P>(service: string, method: string, params: P): Promise<void>;
}

export interface RpcClientCoreOptions {
  defaultTimeoutMs?: number;
}

export class RpcClientCore implements ClamatorClient {
  private state: 'idle' | 'started' | 'stopped' = 'idle';
  private readonly defaultTimeoutMs: number;

  constructor(private readonly transport: Transport, opts: RpcClientCoreOptions = {}) {
    this.defaultTimeoutMs = opts.defaultTimeoutMs ?? 30_000;
  }

  async call<P, R>(service: string, method: string, params: P): Promise<R> {
    if (!SERVICE_RE.test(service)) throw new Error(`invalid service "${service}"`);
    if (!METHOD_RE.test(method)) throw new Error(`invalid method "${method}"`);
    const id = randomUUID();
    const env = buildRequest(`${service}.${method}`, params, id);
    const reply = await this.transport.send(env, { timeoutMs: this.defaultTimeoutMs });
    let parsed;
    try { parsed = parseEnvelope(reply); }
    catch (e) { throw new ClamatorProtocolError(`invalid response envelope: ${(e as Error).message}`); }
    if (parsed.kind === EnvelopeKind.SuccessResponse) return parsed.result as R;
    if (parsed.kind === EnvelopeKind.ErrorResponse) {
      const { code, message, data } = parsed.error;
      throw new RpcError(code, message, data);
    }
    throw new ClamatorProtocolError(`unexpected response kind: ${parsed.kind}`);
  }

  async notify<P>(service: string, method: string, params: P): Promise<void> {
    if (!SERVICE_RE.test(service)) throw new Error(`invalid service "${service}"`);
    if (!METHOD_RE.test(method)) throw new Error(`invalid method "${method}"`);
    await this.transport.notify(buildNotification(`${service}.${method}`, params));
  }

  async start(): Promise<void> {
    if (this.state === 'stopped') throw new Error('client has been stopped');
    if (this.state === 'started') return;
    await this.transport.start();
    this.state = 'started';
  }

  async stop(): Promise<void> {
    if (this.state !== 'started') { this.state = 'stopped'; return; }
    await this.transport.stop();
    this.state = 'stopped';
  }
}

File: py/packages/protocol/src/clamator_protocol/client_core.py
from __future__ import annotations

import uuid
from typing import Any, Protocol

from .envelope import (
    METHOD_RE,
    SERVICE_RE,
    ErrorResponseEnvelope,
    SuccessResponseEnvelope,
    build_notification,
    build_request,
    parse_envelope,
)
from .error import ClamatorProtocolError, RpcError
from .transport import Transport


class ClamatorClient(Protocol):
    async def call(
        self,
        service: str,
        method: str,
        params: Any,
        *,
        timeout_ms: int | None = None,
    ) -> Any: ...
    async def notify(self, service: str, method: str, params: Any) -> None: ...


class RpcClientCore:
    def __init__(self, transport: Transport, *, default_timeout_ms: int = 30_000) -> None:
        self._transport = transport
        self._default_timeout = default_timeout_ms / 1000
        self._state = "idle"

    async def call(
        self,
        service: str,
        method: str,
        params: Any,
        *,
        timeout_ms: int | None = None,
    ) -> Any:
        if not SERVICE_RE.match(service):
            raise ValueError(f"invalid service \"{service}\"")
        if not METHOD_RE.match(method):
            raise ValueError(f"invalid method \"{method}\"")
        rpc_id = str(uuid.uuid4())
        env = build_request(f"{service}.{method}", params, rpc_id)
        timeout = (timeout_ms / 1000) if timeout_ms is not None else self._default_timeout
        reply = await self._transport.send(env, timeout=timeout)
        try:
            parsed = parse_envelope(reply)
        except ValueError as e:
            raise ClamatorProtocolError(f"invalid response envelope: {e}") from e
        if isinstance(parsed, SuccessResponseEnvelope):
            return parsed.result
        if isinstance(parsed, ErrorResponseEnvelope):
            raise RpcError(parsed.error["code"], parsed.error["message"], parsed.error.get("data"))
        raise ClamatorProtocolError(f"unexpected response kind: {parsed.kind}")

    async def notify(self, service: str, method: str, params: Any) -> None:
        if not SERVICE_RE.match(service):
            raise ValueError(f"invalid service \"{service}\"")
        if not METHOD_RE.match(method):
            raise ValueError(f"invalid method \"{method}\"")
        await self._transport.notify(build_notification(f"{service}.{method}", params))

    async def start(self) -> None:
        if self._state == "stopped":
            raise RuntimeError("client has been stopped")
        if self._state == "started":
            return
        await self._transport.start()
        self._state = "started"

    async def stop(self) -> None:
        if self._state != "started":
            self._state = "stopped"
            return
        await self._transport.stop()
        self._state = "stopped"
